File: backend/src/main/java/io/b2mash/b2b/b2bstrawman/projecttemplate/NameTokenResolver.java
package io.b2mash.b2b.b2bstrawman.projecttemplate;

import io.b2mash.b2b.b2bstrawman.customer.Customer;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.format.TextStyle;
import java.util.Locale;
import java.util.Objects;
import org.springframework.stereotype.Component;

/**
 * Resolves name token patterns for project templates and recurring schedules. Uses simple
 * String.replace() substitution — no template engine. If a parameter is null, its token is left
 * unreplaced in the output string.
 */
@Component
public class NameTokenResolver {

  /**
   * Resolves tokens in a name pattern.
   *
   * <p>Supported tokens: {customer}, {month}, {month_short}, {year}, {period_start}, {period_end}
   *
   * @param pattern the name pattern with optional tokens, e.g. "Bookkeeping - {customer} - {month}
   *     {year}"
   * @param customer the customer whose name replaces {customer}; null leaves the token unreplaced
   * @param referenceDate date used for {month}, {month_short}, {year}; null leaves those tokens
   *     unreplaced
   * @param periodStart replaces {period_start} in ISO format; null leaves it unreplaced
   * @param periodEnd replaces {period_end} in ISO format; null leaves it unreplaced
   * @return the resolved name string
   */
  public String resolveNameTokens(
      String pattern,
      Customer customer,
      LocalDate referenceDate,
      LocalDate periodStart,
      LocalDate periodEnd) {
    Objects.requireNonNull(pattern, "pattern must not be null");
    String result = pattern;
    if (customer != null) {
      result = result.replace("{customer}", customer.getName());
    }
    if (referenceDate != null) {
      // Replace {month_short} before {month} for defensive ordering
      result =
          result.replace(
              "{month_short}",
              referenceDate.getMonth().getDisplayName(TextStyle.SHORT, Locale.ENGLISH));
      result =
          result.replace(
              "{month}", referenceDate.getMonth().getDisplayName(TextStyle.FULL, Locale.ENGLISH));
      result = result.replace("{year}", String.valueOf(referenceDate.getYear()));
    }
    if (periodStart != null) {
      result =
          result.replace("{period_start}", periodStart.format(DateTimeFormatter.ISO_LOCAL_DATE));
    }
    if (periodEnd != null) {
      result = result.replace("{period_end}", periodEnd.format(DateTimeFormatter.ISO_LOCAL_DATE));
    }
    return result;
  }
}

File: frontend/lib/name-token-resolver.ts
/**
 * Client-side replica of backend NameTokenResolver.java
 * Resolves name tokens in project template name patterns.
 * No {period_start}/{period_end} — scheduler-only tokens.
 */
export function resolveNameTokens(
  pattern: string,
  customerName?: string,
  referenceDate?: Date
): string {
  let result = pattern;
  const date = referenceDate ?? new Date();

  if (customerName !== undefined) {
    result = result.replace(/\{customer\}/g, customerName);
    result = result.replace(/\{client\}/g, customerName);
  }

  // Full month name in English: "January", "February", etc.
  const monthFull = date.toLocaleDateString("en-US", { month: "long" });
  result = result.replace(/\{month\}/g, monthFull);

  // Short month name: "Jan", "Feb", etc.
  const monthShort = date.toLocaleDateString("en-US", { month: "short" });
  result = result.replace(/\{month_short\}/g, monthShort);

  // 4-digit year
  const year = String(date.getFullYear());
  result = result.replace(/\{year\}/g, year);

  // Strip any remaining unresolved tokens (e.g. {deceased}, {debtor}, {transaction})
  result = result
    .replace(/\{[a-z_]+\}/g, "")
    .replace(/\s{2,}/g, " ")
    .trim();

  return result;
}
